### ISSUE TITLE
Send RPC command from Control Widget to multiple deivces

### ISSUE THREAD

**Issue by openedhardware** [thingsboard/thingsboard#1232]
I have configured a RPC Control Widget and a device which receives the command from the RPC Control widget.
Actually, I created an alias for that device, and assigned this alias to the Control Widget.

But if I use another alias that has multiple devices, any command on the Control Widget is delivered to the **1st** device only.

Is there anyway to send a command to multiple devices?

**Comment by vparomskiy** [comment]
As I know, all default RPC widgets are working only with one device. You need to create a custom RPC widget that will send a command to all devices (that are defined in the alias)

Here is a reference:
https://thingsboard.io/docs/user-guide/contribution/widgets-development/#rpc-control-widget

**Comment by openedhardware** [comment]
@vparomskiy 

Yeah, I created an alias for multiple devices, but the RPC command was sent to the 1st one only...

**Comment by vparomskiy** [comment]
Here is a referecne implementation. It will send RPC command to all devices defined in the alias:

```
var requestBody = {
    method: 'testMethod',
    params: 'someParams'
};

var $scope = self.ctx.$scope;
var deviceService = $scope.$injector.get('deviceService')
var aliasId = self.ctx.defaultSubscription.targetDeviceAliasId;
var aliasInfo = self.ctx.aliasController.getAliasInfo(aliasId);

var deviceIds = [];
for(var i = 0; i<aliasInfo.$$state.value.resolvedEntities.length; i++) {
    var deviceId = aliasInfo.$$state.value.resolvedEntities[i].id;
    deviceIds.push(deviceId);
    deviceService.sendOneWayRpcCommand(deviceId, requestBody).then(
        function success(response) {
            console.log('ok responce');
        },
        function fail(rejection) {
            console.log('reject responce');
        }
        
    );
}
```

**Comment by openedhardware** [comment]
PERFECT!

It works like a charm!

Closing this issue now... :)

**Comment by openedhardware** [comment]
@vparomskiy 

I am getting a couple of errors in this code. (This has been working well til last week, but suddenly stopped working... )

```
Error: aliasInfo is undefined
```

If I use the `device_id` to a known one(so without getting all `device_id` of assigned devices), I am getting this error then:

```
XML Parsing Error: no root element found
Location: https://mytk.ch/api/plugins/rpc/oneway/7b93a780-e18e-11e8-90a1-27e75d0d674e
Line Number 1, Column 1:
```
Any idea?

**Comment by openedhardware** [comment]
NOTE: TB server is hosted on `https://mytk.ch`

**Comment by vparomskiy** [comment]
@openedhardware does this issue still valid for you?

**Comment by alexmed** [comment]
(Thingsboard Community edition, 2.3.0 on AWS EC2 deployed automatically via Doc's link)
most of things working flawlessly

But i think i'm having kind of similar issue but there is important detail:

This part is working:
1. If im using dashboard+device logged as Tenant (no matter if dashboard+device public or not) - everything works ok, particulary Control Widges like "Switch Control" configured to perform RPC call.

This part is not working:
2. if I make dashboard+device Public or assign dashboard+device to some Client/User - the "Switch Control" stops working and Chrome console shows 

"Failed to load resource: the server responded with a status of 403 () /api/plugins/rpc/oneway/<deviceID>"

If i click to the link like my-server/api/plugins/rpc/oneway/165f1d10-400f-11e9-872c-8d390fa314c2
it shows "{"status":401,"message":"Authentication failed","errorCode":10,"timestamp":1551960499467}"

So basically it works in Tenant mode, and not works in User/Public mode

can you please advice how to solve that?

Thanks for great product indeed!

**Comment by ikulikov** [comment]
This is known issue and already fixed in master branch. This fix will be
available in upcoming 2.3.1 release.

**Comment by alexmed** [comment]
thank you, if i look better i could find it in issues list - sorry, now i found lots of this issue mentions

Can you please advice any simple/safe way to upgrade current 2.3.0 to master branch or its better to wait for official 2.3.1 release?

**Comment by buidk17** [comment]
> Here is a referecne implementation. It will send RPC command to all devices defined in the alias:
> 
> ```
> var requestBody = {
>     method: 'testMethod',
>     params: 'someParams'
> };
> 
> var $scope = self.ctx.$scope;
> var deviceService = $scope.$injector.get('deviceService')
> var aliasId = self.ctx.defaultSubscription.targetDeviceAliasId;
> var aliasInfo = self.ctx.aliasController.getAliasInfo(aliasId);
> 
> var deviceIds = [];
> for(var i = 0; i<aliasInfo.$$state.value.resolvedEntities.length; i++) {
>     var deviceId = aliasInfo.$$state.value.resolvedEntities[i].id;
>     deviceIds.push(deviceId);
>     deviceService.sendOneWayRpcCommand(deviceId, requestBody).then(
>         function success(response) {
>             console.log('ok responce');
>         },
>         function fail(rejection) {
>             console.log('reject responce');
>         }
>         
>     );
> }
> ```

Hi,

Can you please explain how to create a widget with this code please ? I am new to ThingsBoard and coding.
Thank you

**Comment by vcmorini** [comment]
> deviceService.sendOneWayRpcCommand

Hello,

How is possible to know that **deviceService** has the **sendOneWayRpcCommand**  method?

**Comment by francisco-hoo** [comment]
Hi guys, I'm using the Thingsboard 3.0.1 but unfortunately I got:
```js
Widget Error: NullInjctorError: R3.Injector(e)[deviceService -> deviceService -> deviceService -> deviceService]: NullInjectorError: No provider for deviceService!
``` 

On line:
```js
var deviceService = $scope.$injector.get('deviceService');
```

Have any suggestions guys?

**Comment by rnandan273** [comment]
With the snippet mentioned above, i tried to replicate and add a custom action on the control widget.
I have  added an action button on the widget to control the group of devices
The alias configured for the widget is a Device Group 

var requestBody = {
    method: ‘set_light_state’,
    params: ‘1’
};
console.log("control widget -> ");
var aliasId = widgetContext.defaultSubscription.targetDeviceAliasId;
console.log("control widget -> "+aliasId);
var aliasInfo = widgetContext.aliasController.getAliasInfo(aliasId);
console.log("AliasInfo"+aliasInfo);
var deviceIds = [];
for(var i = 0; i<aliasInfo.$$state.value.resolvedEntities.length; i++) {
    var deviceId = aliasInfo.$$state.value.resolvedEntities[i].id;
    console.log(deviceId);
widgetContext.deviceService.sendOneWayRpcCommand(aliasId, requestBody).then(
        function success(response) {
            console.log('ok responce');
        },
        function fail(rejection) {
            console.log('reject responce');
        }
    );
}
======================================================================
I am able to get the aliasInfo but not able to resolve the entities in the group.
Any ideas?

**Comment by Skyflowed** [comment]
I finally found a solution :

Create a new custom widget with type : TIME SERIES (it won't work with others types...)

Put the following code into each section :

JS :
self.onInit = function() {
    console.log("NON RPC TYPE");
    console.log(self.ctx);
    self.ctx.$scope.sendCommand = function() {
        var rpcMethod = self.ctx.settings.rpcMethod;
        var selected_command = document.getElementById("command").value;
        console.log(document.getElementById("command").value);
        var rpcParams = selected_command == "on" ? "true" : "false";
        var timeout = self.ctx.settings.requestTimeout;
        var oneWayElseTwoWay = self.ctx.settings.oneWayElseTwoWay ? true : false;
        var requestBody = {"method": rpcMethod, "params": rpcParams};
        var deviceIds = [];
        // Get deviceService 
        var $injector = self.ctx.$scope.$injector;
        var $deviceservice = $injector.get(self.ctx.servicesMap.get('deviceService'));
        for(i=0;i<self.ctx.datasources.length;i++){
            deviceIds.push(self.ctx.datasources[i].entityId);
        }
        deviceIds.forEach(element => {
             var commandObservable;
            if (oneWayElseTwoWay) {
                // Send oneWay RPC-Command, with requestBody like {"method": "methodname", "params": 1} 
                commandObservable  = $deviceservice.sendOneWayRpcCommand(element, requestBody, { ignoreErrors: true  });
            } else {
                commandObservable = self.ctx.controlApi.sendTwoWayCommand(element, requestBody, timeout);
            }
            commandObservable.subscribe(
                function (response) {
                    console.log("response");
                    if (oneWayElseTwoWay) {
                        self.ctx.$scope.rpcCommandResponse = "Command was successfully received by device.<br> No response body because of one way command mode.";
                    } else {
                        self.ctx.$scope.rpcCommandResponse = "Response from device:<br>";                    
                        self.ctx.$scope.rpcCommandResponse += JSON.stringify(response, undefined, 2);
                    }
                    self.ctx.detectChanges();
                },
                function (rejection) {
                    self.ctx.$scope.rpcCommandResponse = "Failed to send command to the device:<br>"
                    self.ctx.$scope.rpcCommandResponse += "Status: " + rejection.status + "<br>";
                    self.ctx.$scope.rpcCommandResponse += "Status text: '" + rejection.statusText + "'";
                    self.ctx.detectChanges();
                }
                
            );
        })
        
       
    }
}

HTML : 

<form #rpcForm="ngForm" (submit)="sendCommand()">
     <center>
        <label for="command">Command type :</label><br>
        <select name="command" id="command">
            <option value="on">ON</option>
            <option value="off">OFF</option>
        </select><br><br>
        <button type = submit>SEND RPC COMMAND TO DEVICE GROUP</button><br><br>
     </center>
        <div>
            <center>
                <label>Command response :</label>
                <div style="width: 100%; height: 100px; border: solid 2px gray" [innerHTML]="rpcCommandResponse">
          </div>
          </center>
        </div>
</form>

SETTINGS SCHEMA : 

{
    "schema": {
        "type": "object",
        "title": "Settings",
        "properties": {
            "oneWayElseTwoWay": {
                "title": "Is One Way Command",
                "type": "boolean",
                "default": true
            },
            "requestTimeout": {
                "title": "RPC request timeout",
                "type": "number",
                "default": 500
            },
            "rpcMethod": {
                "title": "rpcMethod",
                "type" : "string",
                "default" : "setValue"
            }
        },
        "required": []
    },
    "form": [
        "oneWayElseTwoWay",
        "requestTimeout",
        "rpcMethod"
    ]
}

**Comment by parcox** [comment]
Does anybody know how to send RPC command to multiple devices from rule chain? This is my situation, I have a device that defines some parameters. It can trigger LED or buzzer of some other devices, depends on the parameters value.

**Comment by truongkmt** [comment]
I reference above and change some link/directory. it work.
Thank you,

**Comment by Ultrazombie** [comment]
Closed due to inactivity
